src/formats/CML.cpp

```diff
@@ -0,0 +1,581 @@
+                }
+            }
+        }
+        frame.set_cell(UnitCell(a, b, c, alpha, beta, gamma));
+    }
+
+    // Frame level properties
+    auto properties_node = current.child("propertyList");
+    if (properties_node) {
+        for (auto property : properties_node.children("property")) {
+            auto title = property.attribute("title");
+            if (!title) {
+                warning("[CML] Skipping untitled property");
+                continue;
+            }
+
+            auto title_str = title.as_string();
+            auto scalar_node = property.child("scalar");
+            if (!scalar_node) {
+                warning("[CML] {} has no scalar associated with it!", title_str);
+                continue;
+            }
+
+            read_property_(frame, scalar_node, title_str);
+        }
+    }
+
+    // Read the atoms
+    std::unordered_map<std::string, size_t> ref_to_id;
+
+    // TODO: Read string vector versions instead of just the nodes
+    for (auto atom : current.child("atomArray").children("atom")) {
+        double x2, y2, x3, y3, z3, xf, yf, zf;
+        x2 = y2 = x3 = y3 = z3 = xf = yf = zf = 0.0;
+        std::string id, element, title;
+        double formal_charge = 0.0, isotope = 0.0, hydrogen_count = 0.0;
+        for (auto attribute : atom.attributes()) {
+            auto name = std::string(attribute.name());
+            if (name == "id") {
+                id = attribute.as_string();
+            } else if (name == "elementType") {
+                element = attribute.as_string();
+            } else if (name == "x2") {
+                x2 = attribute.as_double();
+            } else if (name == "y2") {
+                y2 = attribute.as_double();
+            } else if (name == "x3") {
+                x3 = attribute.as_double();
+            } else if (name == "y3") {
+                y3 = attribute.as_double();
+            } else if (name == "z3") {
+                z3 = attribute.as_double();
+            } else if (name == "xFract") {
+                xf = attribute.as_double();
+            } else if (name == "yFract") {
+                yf = attribute.as_double();
+            } else if (name == "zFract") {
+                zf = attribute.as_double();
+            } else if (name == "title") { // special attributes
+                title = attribute.as_string();
+            } else if (name == "formalCharge") {
+                formal_charge = attribute.as_double();
+            } else if (name == "isotopeNumber") {
+                isotope = attribute.as_double();
+            } else if (name == "hydrogenCount") {
+                hydrogen_count = attribute.as_double();
+            } else {
+                warning("[CML] Unknown atom attribute: " + name);
+            }
+        }
+
+        auto new_atom = Atom(title, element);
+        Vector3D cart, velo;
+
+        if (crystal_node && (xf != 0.0 || yf != 0.0 || zf != 0.0)) {
```

Comment:
what happens if crystal_node does not exist but xf/yf/zf are set ?

Reply:
xf/yf/zf get ignored.

Reply:
So then x3/y3/z3 get used ? Can both appear in the same file ?

Reply:
Yes, the CML standard is quite loose. You can have x2, y2, x3, y3, z3, xf, yf, and zf all defined for the same file.